### ISSUE TITLE
Enhancement: [DeepPartial] Support for optionally excluding user-provided types. 

### ISSUE THREAD

**Issue by PieterDexper** [ts-essentials/ts-essentials#361]
### What

I propose to add an additional optional generic parameter to DeepPartial (and possibly other similar types) that allows the user to specify their own types that should be excluded from being made (Deep)Partial on top of the already built in exceptions. 
`DeepPartial<Type, Exclude>`

Any nested type matching `Exclude` would not be made recursively partial. 

This is helpful in cases where the system has non-standard built in types that can never be partially present, which currently can't be excluded from DeepPartial without fully re-implementing it. 

### Examples

```
type Example = {
   foo: Timestamp;
}

const a: DeepPartial<Example> = ...
const b: DeepPartial<Example, Timestamp> = ...

const timestampA = a.foo?.toMillis() // Error, toMillis may be undefined.
const timestampA = a.foo?.toMillis?.() // Works, but ugly. 
const timestampB = b.foo?.toMillis() // No error, Timestamp was excluded from DeepPartial. 
```

### Additional Info

I'm unsure if this meets the standards of simplicity and general usability, would this be a welcome addition?

**Comment by Beraliv** [comment]
Hey @PieterDexper 👋

Thanks for the issue!

In your specific example you can simply use `MarkOptional` – https://tsplay.dev/N55DPN

If you want to do it on more than 1 layer, I'm happy to create a separate `DeepMarkOptional`

Will it be enough for your use case?

**Comment by PieterDexper** [comment]
Hi @Beraliv

Thanks for the input! I'm not sure a recursive version of MarkOptional won't result in the same problem. 

A quick check with:
```
MarkOptional<Timestamp, keyof Timestamp>
```
results in its methods giving the same error.

If I understand it correctly with a recursive version of MarkOptional I would be able to use this to exclude specific keys from being made optional, which would be an improvement because I could exclude Timestamp's `toDate` and `toMillis` from the checks recursively, and then ensure they are not in use as key names anywhere else, or alternatively standardize the names of Timestamp fields and exclude those instead.

This isn't ideal because the full setup I have uses DeepPartial within integration tests on data models that describe everything in our system, which results in the issue above with the built in Timestamp type. Since it's built on a document-based database the timestamps can appear within the types at any level of nesting, with arbitrary key names. Those keys may also be in use elsewhere containing a different type. 

My workaround at the moment re-implements DeepPartial with this added to the various checks in there:
```
type SelectiveDeepPartial<Type> = Type extends Timestamp 
    ? Type
    : //rest of checks from DeepPartial 
```

**Comment by Beraliv** [comment]
Currently ts-essentials doesn't have a way to specify what will be excluded and what not.

Currently https://github.com/ts-essentials/ts-essentials/blob/master/lib/deep-partial/index.ts#L5 there are the following types that aren't extended:

* `Builtin` (except for `Error` now)
* `unknown`

> This isn't ideal because the full setup I have uses DeepPartial within integration tests on data models that describe everything in our system, which results in the issue above with the built in Timestamp type

I see it now. Can you make up the simple example that explains what you're doing in tests? That will be important to mention so people will understand why it's designed this way and what use cases it covers.

Speaking of implementation, I think that can be changed with default generic type.

Let's say we will have two generic types now `DeepPartial<Type, NonExtendableMembers = Exclude<Builtin, Error>> = ...` but if you want to extend it, you will be able to change it to `DeepPartial<Type, Exclude<Builtin, Error>> | Timestamp`

I will think whether I want it to be configurable, maybe I will have generic type differently if anything needs to be added in the nearest future, e.g. `DeepPartial<Type, { extendable: Error; nonExtendable: Builtin | Timestamp }`, so backwards compatibility won't be broken

Quick PoC in here – https://tsplay.dev/mZ1x1N

To note, Similarly to `include` and `exclude` in [tsconfig.json](https://www.typescriptlang.org/tsconfig#include), exclude will be applied first, then include

There's another related issue where property key is omitted recursively, you might be interested to have a look at it – https://github.com/ts-essentials/ts-essentials/issues/104

What do you think?

**Comment by WillsterJohnson** [comment]
My two cents on the `nonExtendable` - I think a more obvious name for it would be something like `nonPartial` or more verbosely `excludeFromPartial`. Users can then be fully assured from the name alone that `DeepPartial` will not touch any values with a matching type to the given union. Something like `nonExtendable` is a little too obscure, it doesn't immediately strike me as meaning "these things will be left alone".

**Comment by goldingdamien** [comment]
※ I think this is part of the same issue, if it is not please let me know and I'll make a new issue.
DeepPartial is changing arrays like below:

key?: string[]
=>
key?: (string|undefined)[] | undefined

How can I have string instead of (string|undefined) ?

**Comment by Beraliv** [comment]
> ※ I think this is part of the same issue, if it is not please let me know and I'll make a new issue.
> 
> DeepPartial is changing arrays like below:
> 
> 
> 
> key?: string[]
> 
> =>
> 
> key?: (string|undefined)[] | undefined
> 
> 
> 
> How can I have string instead of (string|undefined) ?
> 
> 

@goldingdamien correct, it is recursively called for all types including arrays so if you'd like to prevent it, there has to be an option to disable it for arrays so arrays stay as is.

I am considering the option that will turn on and off specific types by the names (e.g. date, error, array, etc)

I will keep you in touch when I have a draft version

**Comment by zirkelc** [comment]
> Hey [@PieterDexper](https://github.com/PieterDexper) 👋
> 
> Thanks for the issue!
> 
> In your specific example you can simply use `MarkOptional` – https://tsplay.dev/N55DPN
> 
> If you want to do it on more than 1 layer, I'm happy to create a separate `DeepMarkOptional`
> 
> Will it be enough for your use case?

Hey, I would be interested in a `DeepMarkOptional` type 👍

**Comment by Beraliv** [comment]
Hey @zirkelc! Unfortunately, I don't have capacity to work on it at the moment. If you have time and can raise a PR, I'll direct you and help you out. Let me know if you're interested

**Comment by zirkelc** [comment]
@Beraliv sure I gave it a try: https://github.com/ts-essentials/ts-essentials/pull/432

Let me know if I missed anything!

**Comment by Beraliv** [comment]
@zirkelc I'm happy with your change! @som-sm thank you very much for reviewing the changes. Any objections for me to get it merged and release?

**Comment by Beraliv** [comment]
This will be released as part of https://github.com/ts-essentials/ts-essentials/pull/429. @zirkelc leave it with me, it should be done this weekend

**Comment by Beraliv** [comment]
The change has been released as https://github.com/ts-essentials/ts-essentials/releases/tag/v10.1.0, thank you for your great work, @zirkelc, and for your review @som-sm! cc @PieterDexper
